ironfish/src/syncer.ts

```diff
@@ -99,46 +112,169 @@ export class Syncer {
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    let syncCandidates = ArrayUtils.shuffle(peers)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (currentPeerIdentity) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer) {
+        syncCandidates = syncCandidates.filter((p) => p.state.identity !== currentPeerIdentity)
+        syncCandidates.unshift(currentPeer)
+      }
+    }
+
+    const peerRtt = new Map<Identity, number>()
+
+    // Measure how long it takes to fetch the genesis block header from each
+    // peer as an estimate of connection quality
+    for (const peer of syncCandidates) {
+      if (this.state === 'stopped' || this.state === 'stopping') {
+        return
+      }
+
+      // We only want to successfully measure so many candidates per measurement
+      // phase
+      if (peerRtt.size >= CANDIDATES_PER_MEASUREMENT) {
+        break
+      }
+
+      if (peer.state.type !== 'CONNECTED') {
+        continue
+      }
+
+      const peerIdentity = peer.getIdentityOrThrow()
+
+      const start = BenchUtils.start()
+      try {
+        const response = await this.peerNetwork.getBlockHeaders(peer, 1, 1)
+        if (response.headers.length !== 1) {
+          this.logger.warn(`Peer ${peer.displayName} sent the wrong number of block headers`)
+          peer.punish(BAN_SCORE.MAX, 'invalid response')
+          continue
+        }
+        if (!response.headers[0].hash.equals(this.chain.genesis.hash)) {
+          this.logger.warn(`Peer ${peer.displayName} sent the wrong block header`)
+          peer.punish(BAN_SCORE.MAX, 'invalid response')
+          continue
+        }
+      } catch (e) {
+        this.logger.debug(
+          `Error while trying to measure peer '${
+            peer.displayName
+          }', skipping this peer: ${ErrorUtils.renderError(e)}`,
+        )
+        continue
+      }
+      const rtt = BenchUtils.end(start)
+      peerRtt.set(peerIdentity, rtt)
+    }
+
+    if (peerRtt.size === 0) {
+      return
+    }
+
+    // Sort the peers by the round-trip-time of the block header request and get
+    // the fastest one to sync from
+    const fastestCandidateIdentity = [...peerRtt.entries()].sort((a, b) => a[1] - b[1])[0][0]
+    const peer = this.peerNetwork.peerManager.getPeer(fastestCandidateIdentity)
+
+    const measurementTime = BenchUtils.end(measurementStart)
```

Comment:
Maybe consider bumping this above `if (peerRtt.size === 0) {` so that it still logs the measurementTime even if all the attempts errored out, for some reason (or there weren't any peers). Speaking of which, adding `peerRtt.size` to that log might be helpful too